shared/actions/chat/thread-content.js

```diff
@@ -744,21 +748,65 @@ function* _updateThread({
   }
 }
 
-function* _appendMessagesToConversation({
-  payload: {conversationIDKey, messages},
-}: ChatGen.AppendMessagesPayload) {
-  const currentMessages = yield Saga.select(Constants.getConversationMessages, conversationIDKey)
-  const nextMessages = currentMessages.concat(messages.map(m => m.key))
-  yield Saga.put(
-    EntityCreators.replaceEntity(['conversationMessages'], I.Map({[conversationIDKey]: nextMessages}))
-  )
+function _getMessageOrdinal(m: Constants.ServerMessage): number {
+  switch (m.type) {
+    case 'Attachment':
+    case 'Text':
+      // These pending messages do not have message IDs, instead getting the special ordinal
+      // values that will keep them in the correct order in the message list.
+      if (m.messageState === 'pending' || m.messageState === 'failed') {
+        return m.ordinal
+      }
+      return m.rawMessageID
+    default:
+      return m.rawMessageID
+  }
 }
 
-function* _prependMessagesToConversation({
+function addMessagesToConversation(
+  state: TypedState,
+  conversationIDKey: Constants.ConversationIDKey,
+  messages: Array<Constants.ServerMessage>
+): Constants.ConversationMessages {
+  const currentMessages = Constants.getConversationMessages(state, conversationIDKey)
```

Comment:
my reading is that low is always the ordinal of the first element of the message list, and high is always the ordinal of the last. (Special case when the list is empty.) If so, why have them at all? Am I missing something?

Reply:
I could have been overly conservative here, but I wanted to make as few changes to the usage of the "message key list" as possible. Storing the list inside `ConversationMessages` as a list of `MessageKeys` allowed me to just make a bunch of one line changes where it was used (to pull it from the field as opposed to using it directly). I also didn't want to change the format of the `MessageKey`'s themselves, since I really didn't want to cause any collateral damage.

If people feel strongly enough, I can try and drop `high` and `low`, since they aren't strictly needed. It just made the code slightly easier for me to understand while writing it, and I thought safer to integrate with existing code.

Reply:
In my test this looks really strange since I made the "keys" the same as the ordinal, but in the actual code its a string with fields separated by colons. The pending message keys just have the outbox ID in there, so I would have had to make a change there to get ordinal in there. Another option is I just store the full message in the list, and then reduce it when something needs just the keys.  I thought that might be slower if the user had scrolled way up in the convo, or we start loading more messages on initial load (this might not be a real factor)

Reply:
Ah, I don't have a problem with storing the list as a list of `MessageKeys`, just was confused about the repeated info. I'm fine with leaving it how it is, although perhaps add a comment to the type saying that `high` is the ordinal of the last message, etc.

I guess to get rid of high/low, you'd need to do:
1) Handle an empty conversation list specially (just map the passed-in new messages)
2) Set `low = currentMessages.messages[0].ordinal` (or whatever you need to do for immutable lists), and similarly for `high` at the top of the function, and use that.
3) Change the record type (which now just contains a list) to a list type.

so I'm okay putting that off until needed.